### PR TITLE
Gateway v6

### PR DESCRIPTION
**Pull request by Daniel-Worrall**
Implemented v6 to support regular code on the new gateway. Have not put in much support for the group chat yet. PR mainly for testing and visibility

### CODE REVIEW

File: lib/discordrb/data.rb [hunk 1]
```diff
@@ -964,8 +956,8 @@ class Channel
     # @return [Hash<Integer => OpenStruct>] the channel's permission overwrites
     attr_reader :permission_overwrites
 
-    # @return [true, false] whether or not this channel is a PM channel.
-    def private?
+    # @return [true, false] whether or not this channel has a server.
+    def server?
```
Comment (meew0) [inline lib/discordrb/data.rb]: The name `server?` may be confusing, could you keep it as `private?` (since group DMs are also private) and change the comment to something along the lines of "whether or not this channel is private (DM channel or group DM)"

File: lib/discordrb/cache.rb [hunk 2]
```diff
@@ -188,7 +188,7 @@ def invite(invite)
     # Finds a channel given its name and optionally the name of the server it is in.
     # @param channel_name [String] The channel to search for.
     # @param server_name [String] The server to search for, or `nil` if only the channel should be searched for.
-    # @param type [String, nil] The type of channel to search for (`'text'` or `'voice'`), or `nil` if either type of
+    # @param type [Integer, nil] The type of channel to search for (`0:text, 1:private, 2:voice, 3:group`), or `nil` if any type of
```
Comment (meew0) [inline lib/discordrb/cache.rb]: Minor detail but the "0:text, 1:private, 2:voice, 3:group" is in code formatting here, inconsistent with the other occurrences. Also, could you add a space in between (as in, "0: text")? It looks nicer in my opinion.

File: lib/discordrb/data.rb [hunk 3]
```diff
@@ -935,8 +935,8 @@ class Channel
     # @return [Integer] the type of this channel (0: text, 1: private, 2: voice, 3: group)
     attr_reader :type
 
-    # @return [Recipient, nil] the recipient of the private messages, or nil if this is not a PM channel
-    attr_reader :recipient
+    # @return [Array, nil] An array of recipients of the private messages, or nil if this is not a Private channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: The type should be specified as `Array<Recipient>, nil`, to clarify of what type the entries are.

**Comment by meew0** [comment]
Is there any particular reason why you renamed the LICENSE file?

**Comment by Daniel-Worrall** [comment]
I'm not sure why c9 ide renamed it, but I'll revert that.

**Comment by meew0** [comment]
Things that still need to be updated that I noticed during testing:

- `Channel#update_from`
- private channel handling during READY
- private channel handling in `Message#initialize`
- Creating a private channel (`API#create_private`) now uses a `recipients` array of IDs
- `TypingEvent` currently still relies on singular `recipient`

I may have missed something, feel free to also look yourself for other uses of `recipient` and `type`.

**Comment by meew0** [comment]
Oh, also, the REST API version also needs to be changed to v6 (`https://discordapp.com/api/v6`) so there are no inconsistencies in that regard.

**Comment by Daniel-Worrall** [comment]
Looks like the code works fine with the current implementation of "private channel handling in `Message#initialize`" as the message can only have one recipient and does not affect the array. Working on the rest.

**Comment by meew0** [comment]
The problem is in [this](https://github.com/meew0/discordrb/blob/274390685f83dd45b1911749eabf2e35738e01ce/lib/discordrb/bot.rb#L785) line - the method is now called `recipients`. Instead of changing this, you could also simply add a method `recipient` on channel that returns the first recipient for private channels - both would be fine by me.

**Comment by Daniel-Worrall** [comment]
That would be "private channel handling during READY", not "private channel handling in `Message#initialize`" :D

**Comment by meew0** [comment]
Oh whoops, I misread that... sorry

File: lib/discordrb/api.rb [hunk 4]
```diff
@@ -411,20 +411,60 @@ def resolve_invite(token, invite_code)
...
       content_type: :json
     )
   rescue RestClient::BadRequest
     raise 'Attempted to PM the bot itself!'
   end
 
+  # Create an empty group channel
+  def create_empty_group(token, bot_user_id)
+    request(
+        __method__,
```
Comment (meew0) [inline lib/discordrb/api.rb]: These parameters are indented incorrectly

File: lib/discordrb/api.rb [hunk 5]
```diff
@@ -411,20 +411,60 @@ def resolve_invite(token, invite_code)
...
+        :post,
+        "#{api_base}/users/#{bot_user_id}/channels",
+        { }.to_json,
+        Authorization: token,
+        content_type: :json
+    )
+  end
+
+  # Create a group channel
+  def create_group(token, pm_channel_id, user_id)
```
Comment (meew0) [inline lib/discordrb/api.rb]: The second parameter should be "group_channel_id"

File: lib/discordrb/api.rb [hunk 6]
```diff
@@ -411,20 +411,60 @@ def resolve_invite(token, invite_code)
...
+      content_type: json
+    )
+  rescue RestClient::NoContent
+    raise 'Attempted to create a group channel with the PM channel recipient!'
+  rescue RestClient::Forbidden
+    raise 'Attempted to add a user to group channel without permission!'
+  end
+
+  # Add a user to a Group channel
+  def add_group_user(token, group_channel_id, user_id)
```
Comment (meew0) [inline lib/discordrb/api.rb]: How does this method differ from the above? Both do the exact same thing - add a user to a channel. They could be unified into one method that lets you specify the `recipient_id` parameter separately.

File: lib/discordrb/bot.rb [hunk 7]
```diff
@@ -595,13 +595,14 @@ def update_voice_server(data)
...
       # Handle normal and private channels separately
       if server
         server.channels << channel
         @channels[channel.id] = channel
-      else
-        @private_channels[channel.id] = channel
+      elsif channel.pm?
+        @pm_channels[channel.id] = channel
+      elsif channel.group?
+        @group_channels[channel.id] = channel
```
Comment (meew0) [inline lib/discordrb/bot.rb]: There's no reason to handle group channels separately; they are equivalent to normal channels. The reason PM channels are stored separately is to index (and cache) them by recipient ID, which is unnecessary for group ones.

File: lib/discordrb/bot.rb [hunk 8]
```diff
@@ -622,8 +623,10 @@ def delete_channel(data)
       if server
         @channels.delete(channel.id)
         server.channels.reject! { |c| c.id == channel.id }
-      else
-        @private_channels.delete(channel.id)
+      elsif channel.pm?
+        @pm_channels.delete(channel.recipient.id)
+      elsif channel.group?
```
Comment (meew0) [inline lib/discordrb/bot.rb]: Same here

File: lib/discordrb/cache.rb [hunk 9]
```diff
@@ -112,14 +113,23 @@ def member(server_or_id, user_id)
...
+      response = API.create_pm(token, @profile.id, id)
       channel = Channel.new(JSON.parse(response), self)
-      @private_channels[id] = channel
+      @pm_channels[id] = channel
+    end
+
+    # Returns a group channel for the given channel ID
+    # @param id [Integer] The channel ID.
+    # @return [Channel] The channel for the given ID.
+    def group_channel(id)
```
Comment (meew0) [inline lib/discordrb/cache.rb]: ...and here

File: lib/discordrb/cache.rb [hunk 10]
```diff
@@ -15,7 +15,8 @@ def init_cache
       @servers = {}
 
       @channels = {}
-      @private_channels = {}
+      @pm_channels = {}
+      @group_channels = {}
```
Comment (meew0) [inline lib/discordrb/cache.rb]: ...here...

File: lib/discordrb/cache.rb [hunk 11]
```diff
@@ -112,14 +113,23 @@ def member(server_or_id, user_id)
...
+      return @pm_channels[id] if @pm_channels[id]
 
-      response = API.create_private(token, @profile.id, id)
+      debug("Creating pm channel with user id #{id}")
+      response = API.create_pm(token, @profile.id, id)
       channel = Channel.new(JSON.parse(response), self)
-      @private_channels[id] = channel
+      @pm_channels[id] = channel
+    end
+
```
Comment (meew0) [inline lib/discordrb/cache.rb]: Could you re-add the `private_channel` method as an alias? There's no reason to break existing code here.

File: lib/discordrb/data.rb [hunk 12]
```diff
@@ -980,17 +983,18 @@ def initialize(data, bot, server = nil)
       @position = data['position']
       if pm?
         recipient_user = bot.ensure_user(data['recipients'].first)
-        @recipients = [Recipient.new(recipient_user, self, bot)]
-        @name = @recipients.first.username
+        @recipient = Recipient.new(recipient_user, self, bot)
+        @recipients = [@recipient]
```
Comment (meew0) [inline lib/discordrb/data.rb]: This logic is unnecessarily complicated; for either kind of channels `@recipients` should be the default, with the singular `recipient` being added as a getter method instead of a redundant instance variable. The caching logic for the recipients can be unified, with the only difference between group and PM channels being name and owner_id.

File: lib/discordrb/data.rb [hunk 13]
```diff
@@ -1040,6 +1044,8 @@ def send_message(content, tts = false)
       @bot.send_message(@id, content, tts, @server && @server.id)
     end
 
+    alias_method :send, :send_message
```
Comment (meew0) [inline lib/discordrb/data.rb]: Don't alias `send` unnecessarily - it has a meaning in Ruby itself, and while it's possible to overwrite it with custom logic, there's no reason to do that. It will just lead to unexpected behaviour among users.

File: lib/discordrb/data.rb [hunk 14]
```diff
@@ -1148,6 +1154,7 @@ def update_from(other)
       @topic = other.topic
       @name = other.name
       @recipient = other.recipient
+      @recipients = other.recipients
```
Comment (meew0) [inline lib/discordrb/data.rb]: Same here (`@recipients` over redundant `@recipient`)

**Comment by meew0** [comment]
Also, I'm not sure if you're aware about #203 but you might want to rebase onto/merge master, since you're changing stuff in API and that PR refactored the module quite a bit.

**Comment by Daniel-Worrall** [comment]
Working on that as you wrote that

File: lib/discordrb/api.rb (hunk 5 above)
Comment (Daniel-Worrall) [inline lib/discordrb/api.rb]: This is to create a group from an already existing pm channel. The same way on the client you can go into a PM channel with a friend and on the top right add more friends to make a group channel

File: lib/discordrb/api.rb (hunk 6 above)
Comment (Daniel-Worrall) [inline lib/discordrb/api.rb]: I thought about that but they do actually do different things. One takes an existing PM channel and makes a new group channel by adding an extra recipient, while `add_group_user` adds to existing group channels

File: lib/discordrb/cache.rb (hunk 11 above)
Comment (Daniel-Worrall) [inline lib/discordrb/cache.rb]: Will do. Figured `private_channel` would ambiguously incorrectly reference pm & group channels.

File: lib/discordrb/data.rb (hunk 13 above)
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: This was just me organising something that was previously written L1238

Comment (meew0) [inline lib/discordrb/data.rb]: Oh, I thought you just added it here. Nevermind then

**Comment by Daniel-Worrall** [comment]
To do:

- [ ] Verify/add group channel methods. e.g. kick user/edit channel

- [ ] Add group channel related events + triggers e.g. CHANNEL_RECIPIENT_ADD/CHANNEL_RECIPIENT_REMOVE

File: lib/discordrb/api/channel.rb [hunk 15]
```diff
@@ -226,4 +226,44 @@ def update_user_overrides(token, channel_id, user_id, allow, deny)
...
+      content_type: json
+    )
+  rescue RestClient::NoContent
+    raise 'Attempted to create a group channel with the PM channel recipient!'
+  rescue RestClient::Forbidden
+    raise 'Attempted to add a user to group channel without permission!'
+  end
+
+  # Add a user to a Group channel
+  def add_group_user(token, group_channel_id, user_id)
```
Comment (meew0) [inline lib/discordrb/api/channel.rb]: These two methods can still be unified into one.

File: lib/discordrb/bot.rb [hunk 16]
```diff
@@ -598,13 +598,14 @@ def update_voice_server(data)
     def create_channel(data)
       channel = Channel.new(data, self)
       server = channel.server
-
```
Comment (meew0) [inline lib/discordrb/bot.rb]: Why is this line deleted here?

File: lib/discordrb/data.rb [hunk 17]
```diff
@@ -979,25 +974,35 @@ def mention
...
 
+    # @return [Recipient, nil] A PM Recipient or nil if it isn't a PM channel.
+    def recipient
+      @recipients.first if pm?
+    end
+
     # @!visibility private
     def initialize(data, bot, server = nil)
       @bot = bot
-
```
Comment (meew0) [inline lib/discordrb/data.rb]: Same with this line, also seems unnecessarily removed

File: lib/discordrb/data.rb [hunk 18]
```diff
@@ -979,25 +974,35 @@ def mention
...
       @user_limit = data['user_limit']
       @position = data['position']
-
-      @is_private = data['is_private']
-      if @is_private
-        recipient_user = bot.ensure_user(data['recipient'])
-        @recipient = Recipient.new(recipient_user, self, bot)
-        @name = @recipient.username
+      if pm?
+        recipient_user = bot.ensure_user(data['recipients'].first)
```
Comment (meew0) [inline lib/discordrb/data.rb]: This is still not ideal, the same handling as is used below in group channel handling can be used here as well., since `recipients` is also an array (even if it just has one parameter). The handling will differ in name and owner_id but the caching can be done exactly the same way.

**Comment by meew0** [comment]
Also, I only just realised that you're using the `text` type incorrectly; a text channel is any channel that is not a voice channel (types 0, 1, and 3). What you're looking for instead is a guild/server channel.

**Comment by Daniel-Worrall** [comment]
Current build errors when registering the new events to a bot 
```
bot.channel_recipient_add do |event|
end
```

**Comment by Daniel-Worrall** [comment]
I think this is ready for testing/merge. The only thing I have not included is `MESSAGE_ACK` handling and it's unclear what scenarios these events occur as cases where they used to occur no longer do. Almost as if they are being phased out in favour of `MESSAGE_CREATE`

Let me know

**Comment by meew0** [comment]
Don't handle `ACK`, that has been explicitly discouraged by the devs.

File: lib/discordrb/container.rb [hunk 19]
```diff
@@ -182,6 +182,15 @@ def channel_delete(attributes = {}, &block)
       register_event(ChannelDeleteEvent, attributes, block)
     end
 
+    # Todo
```
Comment (meew0) [inline lib/discordrb/container.rb]: This documentation still needs to be done.

File: lib/discordrb/data.rb [hunk 20]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    # @return [Channel] the Group Channel
+    def add_group_users(user_ids)
+      raise 'Attempted to add a user to a non-group channel!' unless group?
+      user_ids = user_ids.to_a if user_ids.is_a? Integer
+      user_ids.each do |user_id|
+        API::Channel.add_group_user(@bot.token, @id, user_id)
+      end
+      self
+    end
+    alias_method :add_group_user, :add_group_users
```
Comment (meew0) [inline lib/discordrb/data.rb]: Please put a new line in between the method and the alias here

File: lib/discordrb/data.rb [hunk 21]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    # @return [Channel] the Group Channel
+    def remove_group_users(user_ids)
+      raise 'Attempted to remove a user from a non-group channel!' unless group?
+      user_ids = user_ids.to_a if user_ids.is_a? Integer
+      user_ids.each do |user_id|
+        API::Channel.remove_group_user(@bot.token, @id, user_id)
+      end
+      self
+    end
+    alias_method :remove_group_user, :remove_group_users
```
Comment (meew0) [inline lib/discordrb/data.rb]: Same here

File: lib/discordrb/data.rb [hunk 22]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+
+    # Leaves the group
+    def leave_group
+      raise 'Attempted to leave a non-group channel!' unless group?
+      API::Channel.leave_group(@bot.token, @id)
+    end
+    alias_method :leave, :leave_group
```
Comment (meew0) [inline lib/discordrb/data.rb]: and here

File: lib/discordrb/data.rb [hunk 23]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+      self
+    end
+    alias_method :add_group_user, :add_group_users
+
+    # Removes a user from a Group channel
+    # @param user_ids [Array<Integer>, Integer] user ID or Array of user IDs to remove from the group channel.
+    # @return [Channel] the Group Channel
+    def remove_group_users(user_ids)
+      raise 'Attempted to remove a user from a non-group channel!' unless group?
+      user_ids = user_ids.to_a if user_ids.is_a? Integer
```
Comment (meew0) [inline lib/discordrb/data.rb]: This line is wrong; you can't turn an integer into an array with `to_a`. Use `[user_ids]` instead.

File: lib/discordrb/data.rb [hunk 24]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+      channel = Channel.new(JSON.parse(response), @bot)
+      channel.add_group_users(user_ids)
+    end
+
+    # Adds a user to a Group channel
+    # @param user_ids [Array<Integer>, Integer] user ID or Array of user IDs to add to the group channel.
+    # @return [Channel] the Group Channel
+    def add_group_users(user_ids)
+      raise 'Attempted to add a user to a non-group channel!' unless group?
+      user_ids = user_ids.to_a if user_ids.is_a? Integer
```
Comment (meew0) [inline lib/discordrb/data.rb]: This line is also wrong (see below)

File: lib/discordrb/data.rb [hunk 25]
```diff
@@ -1022,12 +1030,22 @@ def initialize(data, bot, server = nil)
...
     end
 
-    # @return [true, false] whether or not this channel is a voice channel
+    # @return [true, false] whether or not this channel is a voice channel.
     def voice?
-      @type == VOICE_TYPE
+      @type == 2
+    end
+
+    # @return [true, false] whether or not this channel is a grouo channel.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "group"

File: lib/discordrb/data.rb [hunk 26]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
     def start_typing
       API::Channel.start_typing(@bot.token, @id)
     end
 
-    alias_method :send, :send_message
-    alias_method :invite, :make_invite
+    # Creates a Group channel
+    # @param user_ids [Array<Integer>] Array of user IDs to add to the new group channel (Excluding
+    # the recipient of the PM channel).
+    # @return [Channel] the created Channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: "the created channel"

File: lib/discordrb/data.rb [hunk 27]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    def create_group(user_ids)
+      raise 'Attempted to create group channel on a non-pm channel!' unless pm?
+      response = API::Channel.create_group(@bot.token, @id, user_ids.shift)
+      channel = Channel.new(JSON.parse(response), @bot)
+      channel.add_group_users(user_ids)
+    end
+
+    # Adds a user to a Group channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: "group channel"

File: lib/discordrb/data.rb [hunk 28]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    # @param user_ids [Array<Integer>, Integer] user ID or Array of user IDs to add to the group channel.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "User ID or array of user IDs"

File: lib/discordrb/data.rb [hunk 29]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    # @return [Channel] the Group Channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: "the group channel"

File: lib/discordrb/data.rb [hunk 30]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+      raise 'Attempted to add a user to a non-group channel!' unless group?
+      user_ids = user_ids.to_a if user_ids.is_a? Integer
+      user_ids.each do |user_id|
+        API::Channel.add_group_user(@bot.token, @id, user_id)
+      end
+      self
+    end
+    alias_method :add_group_user, :add_group_users
+
+    # Removes a user from a Group channel
```
Comment (meew0) [inline lib/discordrb/data.rb]: "group channel"

File: lib/discordrb/data.rb [hunk 31]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
+    # @param user_ids [Array<Integer>, Integer] user ID or Array of user IDs to remove from the group channel.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "User ID or array of user IDs"

File: lib/discordrb/data.rb [hunk 32]
```diff
@@ -1234,21 +1253,87 @@ def make_invite(max_age = 0, max_uses = 0, temporary = false)
...
 
+    # Adds a recipient to a group channel.
+    # @param recipient [Recipient] the recipient to add to the group
+    # @raise [ArgumentError] if tried to add a non-recipient
+    # @note For internal use only
+    # @!visibility private
+    def add_recipient(recipient)
+      raise 'Tried to add recipient to a non-group channel' unless group?
+      raise ArgumentError, 'Tried to add a non-recipient to a group' unless recipient.is_a?(Recipient)
+      API::Channel.add_group_user(@bot.token, @id, recipient.id)
```
Comment (meew0) [inline lib/discordrb/data.rb]: Why does this internal caching method make a REST call? As I see it, it's only used where Discord notifies us that the user has just been added.

File: lib/discordrb/events/channels.rb [hunk 33]
```diff
@@ -118,6 +122,62 @@ def matches?(event)
     end
   end
 
+  # Generic subclass for recipient events (add/remove)
+  class ChannelRecipientEvent < Event
+    # @return [Channel] the channel in question.
+    attr_reader :channel
+    delegate :name, :server, :type, :owner_id, :recipients, :topic, :user_limit, :position, :permission_overwrites, to: :channel
+
+    # @return [Recipient] The recipient that was added/removed from the group
```
Comment (meew0) [inline lib/discordrb/events/channels.rb]: "the recipient"

File: lib/discordrb/api/channel.rb [hunk 34]
```diff
@@ -226,4 +226,65 @@ def update_user_overrides(token, channel_id, user_id, allow, deny)
...
+      Authorization: token,
+      content_type: :json
+    )
+  rescue RestClient::NoContent
+    raise 'Attempted to create a group channel with the PM channel recipient!'
+  rescue RestClient::Forbidden
+    raise 'Attempted to add a user to group channel without permission!'
+  end
+
+  # Add a user to a Group channel
```
Comment (meew0) [inline lib/discordrb/api/channel.rb]: "group channel"

File: lib/discordrb/api/channel.rb [hunk 35]
```diff
@@ -226,4 +226,65 @@ def update_user_overrides(token, channel_id, user_id, allow, deny)
...
+      __method__,
+      :put,
+      "#{Discordrb::API.api_base}/channels/#{group_channel_id}/recipients/#{user_id}",
+      {}.to_json,
+      Authorization: token,
+      content_type: :json
+    )
+  end
+
+  # Remove a user from a Group channel
```
Comment (meew0) [inline lib/discordrb/api/channel.rb]: "group channel"

File: lib/discordrb/bot.rb [hunk 36]
```diff
@@ -779,10 +809,14 @@ def handle_dispatch(type, data)
           ensure_server(element)
         end
 
-        # Add private channels
+        # Add pm and group channels
```
Comment (meew0) [inline lib/discordrb/bot.rb]: "PM and group"

File: lib/discordrb/data.rb [hunk 37]
```diff
@@ -969,35 +964,48 @@ class Channel
     # @return [Hash<Integer => OpenStruct>] the channel's permission overwrites
     attr_reader :permission_overwrites
 
-    # @return [true, false] whether or not this channel is a PM channel.
+    # @return [true, false] whether or not this channel is a PM or Group channel.
```
Comment (meew0) [inline lib/discordrb/data.rb]: "or group"

**Comment by meew0** [comment]
Sorry for all the pedantic comments...

**Comment by meew0** [comment]
Another thing: the `create_group` method, when it is called on a PM channel and the recipient's user ID is in the array, creates a useless `Exception: #<JSON::ParserError: A JSON text must at least contain two octets!>` error. I realise that this use case is invalid, however the error could be more descriptive here.

**Comment by meew0** [comment]
Also, it doesn't seem like you're using API v6 yet, just gateway - I don't think it will change anything anywhere other than the used API base path, however it should still be used for consistency.

File: lib/discordrb/data.rb (hunk 32 above)
Comment (Daniel-Worrall) [inline lib/discordrb/data.rb]: This was me being stupid and forgetting to remove it after testing.

**Comment by Daniel-Worrall** [comment]
Patched up the formatting. You're welcome to the pedantry especially when it comes to documentation. :) I was using API v6 - I think it disappeared in the merge.

As for the `Exception: #<JSON::ParserError: A JSON text must at least contain two octets!>`, I'm unsure how to reproduce that. Let me know if it's still happening and what environment you're using.

**Comment by meew0** [comment]
What I was doing is PM a user on which I was running a userbot, then had the bot eval `event.channel.create_group([66237334693085184])`. (with that being my user ID - replace with your own during testing)

**Comment by Daniel-Worrall** [comment]
Yeah, I'm completely unable to reproduce that, even on older commits. Everything seems complete on my side, just need to review the new changes.

File: lib/discordrb/api/user.rb [hunk 38]
```diff
@@ -95,7 +95,7 @@ def user_dms(token)
 
   # Create a DM to another user
   # https://discordapp.com/developers/docs/resources/user#create-dm
-  def create_private(token, recipient_id)
+  def create_pm(token, recipient_id)
```
Comment (greenbigfrog) [inline lib/discordrb/api/user.rb]: the naming here is a bit wierd. it more like actually creates a DM channel...

**Comment by meew0** [comment]
We could switch to that but that is somewhat outside the scope of this PR. I will look at that after this is merged.

**Comment by greenbigfrog** [comment]
Ok. (I thought it might be sth good to do with the release of 3.0.0, since it might break some things for some people)

**Comment by meew0** [comment]
Well, it might still be 3.0.0, just not this PR.

**Comment by greenbigfrog** [comment]
Yep
